File: java/src/test/java/org/codingdojo/Yatzy2Test.java
package org.codingdojo;

import org.codingdojo.yatzy2.Yatzy2;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;


public class Yatzy2Test {
    Yatzy2 yatzy2 = new Yatzy2();

    @Test
    public void chance_scores_sum_of_all_dice() {
        assertEquals(15, yatzy2.score(List.of(2,3,4,5,1), "CHANCE"));
        assertEquals(16, yatzy2.score(List.of(3,3,4,5,1), "CHANCE"));
    }

    @Test public void yatzy_scores_50() {
        assertEquals(50, yatzy2.score(List.of(4,4,4,4,4), "YATZY"));
        assertEquals(50, yatzy2.score(List.of(6,6,6,6,6), "YATZY"));
        assertEquals(0, yatzy2.score(List.of(6,6,6,6,3), "YATZY"));
    }

    @Test public void test_1s() {
        assertEquals(1, yatzy2.score(List.of(1,2,3,4,5), "ONES"));
        assertEquals(2, yatzy2.score(List.of(1,2,1,4,5), "ONES"));
        assertEquals(0, yatzy2.score(List.of(6,2,2,4,5), "ONES"));
        assertEquals(4, yatzy2.score(List.of(1,2,1,1,1), "ONES"));
    }

    @Test
    public void twos() {
        assertEquals(4, yatzy2.score(List.of(1,2,3,2,6), "TWOS"));
        assertEquals(10, yatzy2.score(List.of(2,2,2,2,2), "TWOS"));
    }

    @Test
    public void threes() {
        assertEquals(6, yatzy2.score(List.of(1,2,3,2,3), "THREES"));
        assertEquals(12, yatzy2.score(List.of(2,3,3,3,3), "THREES"));
    }

    @Test
    public void fours()
    {
        assertEquals(12, yatzy2.score(List.of(4,4,4,5,5), "FOURS"));
        assertEquals(8, yatzy2.score(List.of(4,4,5,5,5), "FOURS"));
        assertEquals(4, yatzy2.score(List.of(4,5,5,5,5), "FOURS"));
    }

    @Test
    public void fives() {
        assertEquals(10, yatzy2.score(List.of(4,4,4,5,5), "FIVES"));
        assertEquals(15, yatzy2.score(List.of(4,4,5,5,5), "FIVES"));
        assertEquals(20, yatzy2.score(List.of(4,5,5,5,5), "FIVES"));
    }
    @Test
    public void sixes() {
        assertEquals(0, yatzy2.score(List.of(4,4,4,5,5), "SIXES"));
        assertEquals(6, yatzy2.score(List.of(4,4,6,5,5), "SIXES"));
        assertEquals(18, yatzy2.score(List.of(6,5,6,6,5), "SIXES"));
    }

    @Test
    public void pair() {
        assertEquals(6, yatzy2.score(List.of(3,4,3,5,6), "PAIR"));
        assertEquals(10, yatzy2.score(List.of(5,3,3,3,5), "PAIR"));
        assertEquals(12, yatzy2.score(List.of(5,3,6,6,5), "PAIR"));
    }

    @Test
    public void two_pair() {
        assertEquals(16, yatzy2.score(List.of(3,3,5,4,5), "TWO_PAIRS"));
        assertEquals(16, yatzy2.score(List.of(3,3,5,5,5), "TWO_PAIRS"));
    }

    @Test
    public void three_of_a_kind()
    {
        assertEquals(9, yatzy2.score(List.of(3,3,3,4,5), "THREE_OF_A_KIND"));
        assertEquals(15, yatzy2.score(List.of(5,3,5,4,5), "THREE_OF_A_KIND"));
        assertEquals(9, yatzy2.score(List.of(3,3,3,3,5), "THREE_OF_A_KIND"));
    }

    @Test
    public void four_of_a_knd() {
        assertEquals(12, yatzy2.score(List.of(3,3,3,3,5), "FOUR_OF_A_KIND"));
        assertEquals(20, yatzy2.score(List.of(5,5,5,4,5), "FOUR_OF_A_KIND"));
        assertEquals(12, yatzy2.score(List.of(3,3,3,3,3), "FOUR_OF_A_KIND"));
    }

    @Test
    public void smallStraight() {
        assertEquals(15, yatzy2.score(List.of(1,2,3,4,5), "SMALL_STRAIGHT"));
        assertEquals(15, yatzy2.score(List.of(2,3,4,5,1), "SMALL_STRAIGHT"));
        assertEquals(0, yatzy2.score(List.of(1,2,2,4,5), "SMALL_STRAIGHT"));
    }

    @Test
    public void largeStraight() {
        assertEquals(20, yatzy2.score(List.of(6,2,3,4,5), "LARGE_STRAIGHT"));
        assertEquals(20, yatzy2.score(List.of(2,3,4,5,6), "LARGE_STRAIGHT"));
        assertEquals(0, yatzy2.score(List.of(1,2,2,4,5), "LARGE_STRAIGHT"));
    }
    @Test
    public void fullHouse() {
        assertEquals(18, yatzy2.score(List.of(6,2,2,2,6), "FULL_HOUSE"));
        assertEquals(0, yatzy2.score(List.of(2,3,4,5,6), "FULL_HOUSE"));
    }

}

File: csharp/Yatzy.Test/Yatzy2Test.cs
using System.Collections.Generic;
using Xunit;

namespace Yatzy.Test;

public class Yatzy2Test
{
    private Yatzy2.Yatzy2 calculator = new Yatzy2.Yatzy2();

    [Fact]
    public void chance_scores_sum_of_all_dice()
    {
        Assert.Equal(15, calculator.Score(new List<int> { 2, 3, 4, 5, 1 }, "CHANCE"));
        Assert.Equal(16, calculator.Score(new List<int> { 3, 3, 4, 5, 1 }, "CHANCE"));
    }

    [Fact]
    public void yatzy_scores_50()
    {
        Assert.Equal(50, calculator.Score(new List<int> { 4, 4, 4, 4, 4 }, "YATZY"));
        Assert.Equal(50, calculator.Score(new List<int> { 6, 6, 6, 6, 6 }, "YATZY"));
        Assert.Equal(0, calculator.Score(new List<int> { 6, 6, 6, 6, 3 }, "YATZY"));
    }

    [Fact]
    public void test_1s()
    {
        Assert.Equal(1, calculator.Score(new List<int> { 1, 2, 3, 4, 5 }, "ONES"));
        Assert.Equal(2, calculator.Score(new List<int> { 1, 2, 1, 4, 5 }, "ONES"));
        Assert.Equal(0, calculator.Score(new List<int> { 6, 2, 2, 4, 5 }, "ONES"));
        Assert.Equal(4, calculator.Score(new List<int> { 1, 2, 1, 1, 1 }, "ONES"));
    }

    [Fact]
    public void twos()
    {
        Assert.Equal(4, calculator.Score(new List<int> { 1, 2, 3, 2, 6 }, "TWOS"));
        Assert.Equal(10, calculator.Score(new List<int> { 2, 2, 2, 2, 2 }, "TWOS"));
    }

    [Fact]
    public void threes()
    {
        Assert.Equal(6, calculator.Score(new List<int> { 1, 2, 3, 2, 3 }, "THREES"));
        Assert.Equal(12, calculator.Score(new List<int> { 2, 3, 3, 3, 3 }, "THREES"));
    }

    [Fact]
    public void fours()
    {
        Assert.Equal(12, calculator.Score(new List<int> { 4, 4, 4, 5, 5 }, "FOURS"));
        Assert.Equal(8, calculator.Score(new List<int> { 4, 4, 5, 5, 5 }, "FOURS"));
        Assert.Equal(4, calculator.Score(new List<int> { 4, 5, 5, 5, 5 }, "FOURS"));
    }

    [Fact]
    public void fives()
    {
        Assert.Equal(10, calculator.Score(new List<int> { 4, 4, 4, 5, 5 }, "FIVES"));
        Assert.Equal(15, calculator.Score(new List<int> { 4, 4, 5, 5, 5 }, "FIVES"));
        Assert.Equal(20, calculator.Score(new List<int> { 4, 5, 5, 5, 5 }, "FIVES"));
    }

    [Fact]
    public void sixes()
    {
        Assert.Equal(0, calculator.Score(new List<int> { 4, 4, 4, 5, 5 }, "SIXES"));
        Assert.Equal(6, calculator.Score(new List<int> { 4, 4, 6, 5, 5 }, "SIXES"));
        Assert.Equal(18, calculator.Score(new List<int> { 6, 5, 6, 6, 5 }, "SIXES"));
    }

    [Fact]
    public void pair()
    {
        Assert.Equal(6, calculator.Score(new List<int> { 3, 4, 3, 5, 6 }, "PAIR"));
        Assert.Equal(10, calculator.Score(new List<int> { 5, 3, 3, 3, 5 }, "PAIR"));
        Assert.Equal(12, calculator.Score(new List<int> { 5, 3, 6, 6, 5 }, "PAIR"));
    }

    [Fact]
    public void two_pair()
    {
        Assert.Equal(16, calculator.Score(new List<int> { 3, 3, 5, 4, 5 }, "TWO_PAIRS"));
        Assert.Equal(16, calculator.Score(new List<int> { 3, 3, 5, 5, 5 }, "TWO_PAIRS"));
    }

    [Fact]
    public void three_of_a_kind()
    {
        Assert.Equal(9, calculator.Score(new List<int> { 3, 3, 3, 4, 5 }, "THREE_OF_A_KIND"));
        Assert.Equal(15, calculator.Score(new List<int> { 5, 3, 5, 4, 5 }, "THREE_OF_A_KIND"));
        Assert.Equal(9, calculator.Score(new List<int> { 3, 3, 3, 3, 5 }, "THREE_OF_A_KIND"));
    }

    [Fact]
    public void four_of_a_knd()
    {
        Assert.Equal(12, calculator.Score(new List<int> { 3, 3, 3, 3, 5 }, "FOUR_OF_A_KIND"));
        Assert.Equal(20, calculator.Score(new List<int> { 5, 5, 5, 4, 5 }, "FOUR_OF_A_KIND"));
        Assert.Equal(12, calculator.Score(new List<int> { 3, 3, 3, 3, 3 }, "FOUR_OF_A_KIND"));
    }

    [Fact]
    public void smallStraight()
    {
        Assert.Equal(15, calculator.Score(new List<int> { 1, 2, 3, 4, 5 }, "SMALL_STRAIGHT"));
        Assert.Equal(15, calculator.Score(new List<int> { 2, 3, 4, 5, 1 }, "SMALL_STRAIGHT"));
        Assert.Equal(0, calculator.Score(new List<int> { 1, 2, 2, 4, 5 }, "SMALL_STRAIGHT"));
    }

    [Fact]
    public void largeStraight()
    {
        Assert.Equal(20, calculator.Score(new List<int> { 6, 2, 3, 4, 5 }, "LARGE_STRAIGHT"));
        Assert.Equal(20, calculator.Score(new List<int> { 2, 3, 4, 5, 6 }, "LARGE_STRAIGHT"));
        Assert.Equal(0, calculator.Score(new List<int> { 1, 2, 2, 4, 5 }, "LARGE_STRAIGHT"));
    }

    [Fact]
    public void fullHouse()
    {
        Assert.Equal(18, calculator.Score(new List<int> { 6, 2, 2, 2, 6 }, "FULL_HOUSE"));
        Assert.Equal(0, calculator.Score(new List<int> { 2, 3, 4, 5, 6 }, "FULL_HOUSE"));
    }
}